grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
+        let mut run_updates = StatusUpdates::default();
+        debug!("running program");
+
+        let executable = ws.join(&self.executable);
+        if !executable.exists() {
+            return Err(anyhow!(
+                "Could not find student executable at {:?}",
+                executable.display()
+            ));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                if !ws.join(&supp).exists() {
+                    return Err(anyhow!("Could not find valgrind suppressions at {}", supp));
+                }
+            }
+        }
+
+        let cmd = self.gen_cmd(ws);
+        sect.add_content(("Run Command", format!("{}", cmd).code()));
+
+        let res = cmd.run_with(&self.executor).await?;
+
+        let log = self.read_logfile(ws)?;
+
+        if !res.status.completed() {
+            match res.status {
+                ExitStatus::Timeout(to) => {
+                    run_updates.add_update(
+                        Update::new_fail("running valgrind", PointQuantity::FullPoints).notes(
+                            format!(
+                                "Your submission timed out after {} seconds :(",
+                                to.as_secs()
+                            ),
+                        ),
+                    );
+                    sect.add_content(run_updates);
+                }
+                ExitStatus::Signal(sig) => match sig {
+                    SignalType::SegFault => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGSEGV!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                    SignalType::Abort => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGABRT!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                },
+                _ => panic!("Expected either Timeout or Signal if command was not completed"),
+            }
+            return Ok(StageResult::new(
+                StageStatus::UnrecoverableFailure,
```

Comment:
Any errors here are recoverable. Valgrind stage is run secondary to run stage, and are used to only catch memory issues. we can still run other stages if the valgrind stage fails.

Reply:
Ah okay sounds good! Will update